File: 1125LunBo/1125LunBo/ViewController.m
//
//  ViewController.m
//  1125LunBo
//
//  Created by WY on 15/11/25.
//  Copyright © 2015年 WY. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()<UICollectionViewDelegate,UICollectionViewDataSource>
@property (weak, nonatomic) IBOutlet UICollectionView *myCollectionView;
@property(nonatomic,strong)NSArray * imgArr ;
@property(nonatomic,strong)NSIndexPath * indexPath ;
@property(nonatomic,assign)NSInteger  currentItem ;
@property(nonatomic,strong)NSTimer * time ;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
     NSLog(@"----%@",self.imgArr);
    UICollectionViewLayout * layout = [[UICollectionViewLayout alloc]init];
    [self.myCollectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"cell"];
    [self.myCollectionView scrollToItemAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:1] atScrollPosition:UICollectionViewScrollPositionLeft animated:NO];
    // Do any additional setup after loading the view, typically from a nib.
    
    [self addTimer];
}
-(void)addTimer
{
    NSTimer * time = [NSTimer timerWithTimeInterval:1.0 target:self selector:@selector(nextImage) userInfo:nil repeats:YES];
    self.time=time;
    [[NSRunLoop mainRunLoop ] addTimer:time forMode:NSRunLoopCommonModes];
}
-(void)removeTimer
{
    [self.time invalidate];
    self.time=nil;
}
-(void )nextImage{
    NSIndexPath *currentIndexPath = [[self.myCollectionView indexPathsForVisibleItems] lastObject];
    // 马上显示回最中间那组的数据
    NSIndexPath *currentIndexPathReset = [NSIndexPath indexPathForItem:currentIndexPath.item inSection:self.imgArr.count/2];
    [self.myCollectionView scrollToItemAtIndexPath:currentIndexPathReset atScrollPosition:UICollectionViewScrollPositionLeft animated:NO];
    NSInteger item  = currentIndexPathReset.item + 1 ;
    NSInteger section = currentIndexPathReset.section;
//     NSLog(@"----%d",section);
    if (item==self.imgArr.count) {
        section ++ ;
        item = 0 ;
    }
    self.currentItem = item ;
        [self.myCollectionView scrollToItemAtIndexPath:[NSIndexPath indexPathForItem:item inSection:section] atScrollPosition:UICollectionViewScrollPositionNone animated:YES];

}
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView{
    [self removeTimer];
}

// called on finger up if the user dragged. decelerate is true if it will continue moving afterwards
- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate{
    
     NSLog(@"----%@",@(decelerate));
    [self addTimer];
}


-(NSString * )imgArr {
    if(_imgArr==nil){
        NSMutableArray * arrM = [NSMutableArray array];
        _imgArr = [[NSArray alloc]init];
        for (int i =1; i < 3; i++) {
            NSString * tempStr = [NSString stringWithFormat:@"minion_%02d",i];
            [arrM addObject:tempStr];
        }
        _imgArr = arrM;
        
    }
    return _imgArr;
}

-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return 30;
}
-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.imgArr.count;
}
-(UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    UICollectionViewCell * cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
    [cell.contentView addSubview: [[UIImageView alloc]initWithImage:[UIImage imageNamed:self.imgArr[indexPath.item]]]];
    self.indexPath = indexPath ;
    return cell;
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView{
    NSInteger item =(NSInteger) (self.myCollectionView.contentOffset.x/self.myCollectionView.bounds.size.width) % self.imgArr.count;
    [self.myCollectionView scrollToItemAtIndexPath:[NSIndexPath indexPathForRow:item inSection:1] atScrollPosition:UICollectionViewScrollPositionNone animated:NO];

}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: FirstDemo/FirstDemo/ViewController.swift
//
//  ViewController.swift
//  FirstDemo
//
//  Created by WY on 15/11/25.
//  Copyright © 2015年 WY. All rights reserved.
//

import UIKit

class ViewController: UITableViewController {

    var persons = [Person]()
    override func viewDidLoad() {
        super.viewDidLoad()
//        let arr = [["name" : "格奥尔" , "age" : "80"],["name" : "格奥尔" , "age" : "80"],["name" : "格奥尔" , "age" : "80"],["name" : "格奥尔" , "age" : "80"],["name" : "格奥尔" , "age" : "80"]]
//        for ddd in arr {
//        
//            let p : Person = Person.init(dic : ddd)
//            print(p)
//            persons.append(p)
//        }

        self.loadData { (p) -> () in
            self.persons = p//在闭包里使用成员变量一定要加self.
            print(self.persons)
            self.tableView.reloadData()
        }
        // Do any additional setup after loading the view, typically from a nib.
    }
    func loadData (finishBlockCallBack : ([Person]) -> () ){
        dispatch_async(dispatch_get_global_queue(0, 0)) { () -> Void in
            
            var tempArr : [Person] = [Person]()
            for _ in 0...25 {
                let name : String = "马德里"
                let age : String = "\(random()%30)"
//                let p = Person()
//                p.name = name
//                p.age = age
                let dic : [String : AnyObject] = ["name" : name , "age" : age]
                let p = Person.init(dic: dic)
                
                tempArr.append(p)
            }
            
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                finishBlockCallBack (tempArr)

            })
            
        }
        
        
        
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }



}
extension ViewController {


    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return persons.count
    }
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell  = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath) as! PersonCell
        cell.person = persons[indexPath.row]
        
        return cell
    }



}
